zingolib/src/wallet/transactions/recording.rs

```diff
@@ -446,16 +293,172 @@ impl TxMapAndMaybeTrees {
+        self.transaction_records_by_id
+            .check_notes_mark_change(&spending_txid);
+
+        Ok(())
+    }
+
+    // Will mark a note as having been spent at the supplied height and spent_txid.
+    /// witness tree requirement:
+    /// remove_witness_mark ~-> note_datum.witnessed_position
+    pub fn mark_note_as_spent<D: DomainWalletExt>(
+        &mut self,
+        spent_nullifier: <D::WalletNote as ShieldedNoteInterface>::Nullifier,
+        spending_txid: TxId,
+        status: ConfirmationStatus,
+        source_txid: TxId,
+        output_index: Option<u32>,
+    ) -> ZingoLibResult<u64>
+    where
+        <D as Domain>::Note: PartialEq + Clone,
+        <D as Domain>::Recipient: traits::Recipient,
+    {
+        Ok(if let Some(height) = status.get_confirmed_height() {
+            // ie remove_witness_mark_sapling or _orchard
+            self.remove_witness_mark::<D>(height, spending_txid, source_txid, output_index)?;
+            if let Some(transaction_spent_from) =
+                self.transaction_records_by_id.get_mut(&source_txid)
+            {
+                if let Some(confirmed_spent_note) = D::to_notes_vec_mut(transaction_spent_from)
+                    .iter_mut()
+                    .find(|note| note.nullifier() == Some(spent_nullifier))
+                {
+                    *confirmed_spent_note.spent_mut() = Some((spending_txid, height.into()));
+                    *confirmed_spent_note.pending_spent_mut() = None;
+
+                    confirmed_spent_note.value()
+                } else {
+                    ZingoLibError::NoSuchNullifierInTx(spending_txid).handle()?
+                }
+            } else {
+                ZingoLibError::NoSuchTxId(spending_txid).handle()?
+            }
+        } else if let Some(height) = status.get_broadcast_height() {
+            // Mark the unconfirmed_spent. Confirmed spends are already handled in update_notes
+            if let Some(transaction_spent_from) =
+                self.transaction_records_by_id.get_mut(&source_txid)
+            {
+                if let Some(unconfirmed_spent_note) = D::to_notes_vec_mut(transaction_spent_from)
+                    .iter_mut()
+                    .find(|note| note.nullifier() == Some(spent_nullifier))
+                {
+                    *unconfirmed_spent_note.pending_spent_mut() =
+                        Some((spending_txid, u32::from(height)));
+                    unconfirmed_spent_note.value()
+                } else {
+                    ZingoLibError::NoSuchNullifierInTx(spending_txid).handle()?
+                }
+            } else {
+                ZingoLibError::NoSuchTxId(spending_txid).handle()?
+            }
+        } else {
+            ZingoLibError::UnknownError.handle()?
+        }) // todO add special error variant
     }
 }
 
 // shardtree
-impl TxMapAndMaybeTrees {
+impl super::TxMapAndMaybeTrees {
```

Comment:
I believe we're moving in favor of absolute paths over super. This is more verbose, is is clearer?
```suggestion
impl crate::wallet::transactions::TxMapAndMaybeTrees {
```